platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,231 @@
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        // update the roster state
+        final RosterState currentRosterState = rosterStateOrThrow();
+        final Bytes candidateRosterHash = RosterUtils.hash(candidateRoster).getBytes();
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(candidateRosterHash)
+                .roundRosterPairs(currentRosterState.roundRosterPairs());
+        this.rosterState.put(rosterStateBuilder.build());
+
+        // update the roster map and commit the changes
+        this.rosterMap.put(ProtoBytes.newBuilder().value(candidateRosterHash).build(), candidateRoster);
+        commit();
+    }
+
+    /**
+     * Determines the initial active roster based on the given software version and initial state.
+     * The active roster is obtained by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the active roster is retrieved from the state or an exception is thrown if not found.
+     *
+     * @param version the software version of the current node
+     * @param initialState the initial state of the platform
+     * @return the active roster which will be used by the platform
+     */
+    @NonNull
+    @Override
+    public Roster determineActiveRoster(
+            @NonNull final SoftwareVersion version, @NonNull final ReservedSignedState initialState) {
+        final boolean softwareUpgrade = detectSoftwareUpgrade(version, initialState.get());
+
+        if (!softwareUpgrade) {
+            final Roster lastUsedActiveRoster = rosterStateAccessor.getActiveRoster();
+            // not in software upgrade mode (i.e., normal restart), return the
+            // active roster present in the state or throw if not found.
+            return Objects.requireNonNull(
+                    lastUsedActiveRoster, "Active Roster must be present in the state during normal network restart.");
+        }
+
+        // software upgrade is detected, we adopt the candidate roster
+        final Roster candidateRoster = Objects.requireNonNull(
+                rosterStateAccessor.getCandidateRoster(),
+                "Candidate Roster must be present in the state before attempting to adopt a roster.");
+        adoptCandidateRoster(initialState.get().getRound() + 1);
+        commit();
+        return candidateRoster;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @NonNull
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * Adopts the candidate roster present in the state as the active roster.
+     * Until the Dynamic Address Book is implemented, this method will be called unconditionally on network upgrade.
```

Comment:
TSS will call this conditionally upon network upgrade.

Reply:
removed.